### PR TITLE
feat: add PinRequest to track CIDs to replicate

### PR DESCRIPTION
**Pull request by olizilla**
Adds:
- PinRequest type to graphql schema. Just a CID, the Content, and how many times we tried to pin it.
- createPinRequest to create a new one or adopt an existing one.
- incrementPinRequestAttempts to +1 the attempts
- tweak importCar to create a PinRequest when it creates a new Content. It has all it needs, so just do it inline.
- add index on created
- add findAllPinRequests that gets you all of them, paginated, oldest first
- add cron job to fetch a batch of 600 PinRequests and pin them to Piñata. Either it pins it and deletes the PinRequest or it fails and increments the attempts count.

License: (Apache-2.0 AND MIT)
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

**Comment by olizilla** [comment]
This should also add some Pin to the pins array on a Content!

**Comment by olizilla** [comment]
next steps
- query the pinning service api on pinata to find the pin status and delegates for a new pin.
- verify that the nodes listed as delegates are the ones that host the pin... not just the entire cluster... otherwise we'll show a pin per node even if the replication count is lower than the cluster node count.
- create Pin and PinLocation objects for each delegate
- create a cron job to periodically recheck the pinata pin status for each cid.

**Comment by olizilla** [comment]
This PR is the absolute MVP of backing up to pinata. I think it makes sense to review / merge this at this point and we can punt the rest to a subsequent iteration.

**Comment by olizilla** [comment]
Example response from Piñata shows 7 delegates, but the cluster replication config is 3 of 6, so we probably can't use that information directly.
```json
{
  "count": 1,
  "results": [
    {
      "requestid": "60b56ced-45de-45fe-8610-baad09667d0b",
      "status": "pinned",
      "created": "2021-07-22T10:01:41.021Z",
      "pin": {
        "cid": "QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR",
        "origins": [],
        "meta": {},
        "name": "lidel.jpg"
      },
      "delegates": [
        "/ip4/143.244.156.201/tcp/4001/p2p/QmakMPrBw9RtJhVH9M3hmgqUe7uCTXfrvXCvCci9UkARKZ",
        "/ip4/188.166.226.146/tcp/4001/p2p/QmTxSm9GJfnqJDkYhJnz2G5jiVsVBu8rxV2x2zhvKfcYji",
        "/ip4/167.99.137.38/tcp/4001/p2p/QmWS845wqn5M59NsL3ozin7Rm25npcATfhGdko9YVxMi9R",
        "/ip4/207.154.249.153/tcp/4001/p2p/QmQEmFgTASSpQ8c2JqCLBvjiWxxzPPdUh2tHHFzeZWtHs9",
        "/ip4/157.245.132.137/tcp/4001/p2p/12D3KooWHtrWp115pAAbqQqNwpyNViKrhX89jis5oHoQu4pvNyks",
        "/ip4/143.198.212.250/tcp/4001/p2p/QmaXcTT9FTjVPZ8PYszxnqJ3WY1CXjXe69VWxUwJomAEWX",
        "/ip4/159.89.95.230/tcp/4001/p2p/QmcZEf6qeoUeHCyuy6zk4FYpV8UXRNceQkjtJyxKzaXbVv"
      ],
      "info": {}
    }
  ]
}
```

File: packages/db/fauna/resources/Function/createPinRequest.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+const name = 'createPinRequest'
+const body = Query(
+  Lambda(
+    ['data'],
+    Let(
+      {
+        cid: Select('cid', Var('data')),
+        content: Call('findContentByCid', Var('cid')),
+        contentRef: Select('ref', Var('content')),
```
Comment (alanshaw) [inline packages/db/fauna/resources/Function/createPinRequest.js:34]: Can return `null` if not found.

```suggestion
        contentRef: If(
          IsNull(Var('content')),
          Abort('content not found'),
          Select('ref', Var('content'))
        )
```

So, probably can remove the `Not(Exists(Var('contentRef')))` later.

File: packages/db/fauna/resources/Function/incrementPinRequestAttempts.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  Select
+} = fauna
+
+const name = 'incrementPinRequestAttempts'
+const body = Query(
+  Lambda(
+    ['id'],
+    Let(
+      {
+        pinRequestRef: Ref(Collection('PinRequests'), Var('id')),
```
Comment (alanshaw) [inline packages/db/fauna/resources/Function/incrementPinRequestAttempts.js:26]:
```suggestion
        pinRequestRef: Ref(Collection('PinRequest'), Var('id')),
```

**Review by alanshaw — CHANGES_REQUESTED** [review]

File: packages/db/fauna/resources/Function/createPinRequest.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        cid: Select('cid', Var('data')),
+        content: Call('findContentByCid', Var('cid')),
+        contentRef: If(
+          IsNull(Var('content')),
+          Abort('content not found'),
+          Select('ref', Var('content'))
+        ),
+        attempts: Select('attempts', Var('data'), null)
+      },
+      Let(
```
Comment (alanshaw) [inline packages/db/fauna/resources/Function/createPinRequest.js:40]: Note: double `Let`

**Review by alanshaw — APPROVED** [review]

File: packages/db/fauna/resources/Function/createPinRequest.js (hunk 3 above)
Comment (olizilla) [reply packages/db/fauna/resources/Function/createPinRequest.js:40]: oh my. nice tho, it simplifies!

**Comment by olizilla** [comment]
_proof of it blend_
```console
$ DEBUG="pinata" npm run start:pinata -w packages/cron

> @web3-storage/cron@0.0.0 start:pinata
> node src/bin/pinata.js

  pinata 📡 Fetcing Pin Requests from DB +0ms
  pinata 📥 Processing 1 Pin Requests +559ms
  pinata 📌 bafkreiabltrd5zm73pvi7plq25pef3hm7jxhbi3kv4hapegrkfpkqtkbme 1/1 pinned on Pinata!  +891ms
  pinata 🎉 Done! Pinned 1 of 1 +1ms
```

**Comment by olizilla** [comment]
_and the fail path is blend too_
```console
$ DEBUG="*" npm run start:pinata -w packages/cron                                                                                                                     16:03:12

> @web3-storage/cron@0.0.0 start:pinata
> node src/bin/pinata.js

  pinata 📡 Fetcing Pin Requests from DB +0ms
  pinata 📥 Processing 1 Pin Requests +466ms
  fetchJSON 💥 fetch https://api.pinata.cloud/pinning/pinByHash Error: 401 Unauthorized: {"error":{"reason":"INVALID_CREDENTIALS","details":"Invalid/expired credentials"}}
    at file:///Users/oli/Code/web3-storage/web3.storage/packages/cron/src/lib/fetch.js:32:13
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RetryOperation._fn (/Users/oli/Code/web3-storage/web3.storage/node_modules/p-retry/index.js:50:12) {
  response: Response {
    size: 0,
    timeout: 0,
    [Symbol(Body internals)]: { body: [PassThrough], disturbed: true, error: null },
    [Symbol(Response internals)]: {
      url: 'https://api.pinata.cloud/pinning/pinByHash',
      status: 401,
      statusText: 'Unauthorized',
      headers: [Headers],
      counter: 0
    }
  },
  attemptNumber: 1,
  retriesLeft: 1
} +0ms
  fetchJSON 💥 fetch https://api.pinata.cloud/pinning/pinByHash Error: 401 Unauthorized: {"error":{"reason":"INVALID_CREDENTIALS","details":"Invalid/expired credentials"}}
    at file:///Users/oli/Code/web3-storage/web3.storage/packages/cron/src/lib/fetch.js:32:13
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RetryOperation._fn (/Users/oli/Code/web3-storage/web3.storage/node_modules/p-retry/index.js:50:12) {
  response: Response {
    size: 0,
    timeout: 0,
    [Symbol(Body internals)]: { body: [PassThrough], disturbed: true, error: null },
    [Symbol(Response internals)]: {
      url: 'https://api.pinata.cloud/pinning/pinByHash',
      status: 401,
      statusText: 'Unauthorized',
      headers: [Headers],
      counter: 0
    }
  },
  attemptNumber: 2,
  retriesLeft: 0
} +5s
  pinata 💥 bafkreih4nwkd3lo5sx4qvnligztznyrghgz3bfhleojxfequ7sio6x7se4 1/1 failed to pin. 97 tries left Error: 401 Unauthorized: {"error":{"reason":"INVALID_CREDENTIALS","details":"Invalid/expired credentials"}}
    at file:///Users/oli/Code/web3-storage/web3.storage/packages/cron/src/lib/fetch.js:32:13
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RetryOperation._fn (/Users/oli/Code/web3-storage/web3.storage/node_modules/p-retry/index.js:50:12) {
  response: Response {
    size: 0,
    timeout: 0,
    [Symbol(Body internals)]: { body: [PassThrough], disturbed: true, error: null },
    [Symbol(Response internals)]: {
      url: 'https://api.pinata.cloud/pinning/pinByHash',
      status: 401,
      statusText: 'Unauthorized',
      headers: [Headers],
      counter: 0
    }
  },
  attemptNumber: 2,
  retriesLeft: 0
} +5s
  pinata 🎉 Done! Pinned 0 of 1 +1ms
```

File: packages/cron/src/lib/fetch.js [hunk 4]
```diff
@@ -5,9 +5,9 @@ import debug from 'debug'
 
 const log = debug('fetchJSON')
 
-const REQUEST_TIMEOUT = 60000
-const RETRY_INTERVAL = 60000
-const RETRY_ATTEMPTS = 5
+const REQUEST_TIMEOUT = 10000
+const RETRY_INTERVAL = 5000
+const RETRY_ATTEMPTS = 3
```
Comment (olizilla) [inline packages/cron/src/lib/fetch.js:10]: @alanshaw dialling downt the retry logic on the pianta cron for an individual fail... we can always pick it up again on the next pass.

**Pull request merged**